source/grid/grid_in.cc

```diff
@@ -3252,6 +3252,88 @@ GridIn<dim, spacedim>::read_partitioned_msh(const std::string &file_prefix,
           coords[3 * i + d];
     }
 
+  // --- Collect physical groups for boundary and manifold info ---
+  std::map<std::pair<int, int>, types::boundary_id> entity_to_boundary;
+  std::map<std::pair<int, int>, types::material_id> entity_to_material;
+
+
+  std::vector<std::pair<int, int>> physical_groups;
+  gmsh::model::getPhysicalGroups(physical_groups);
+
+  for (const auto &pg : physical_groups)
+    {
+      int physical_dim = pg.first;
+      int physical_tag = pg.second;
+
+      std::vector<int> physical_entities;
+      gmsh::model::getEntitiesForPhysicalGroup(physical_dim,
+                                               physical_tag,
+                                               physical_entities);
+
+      for (int ent : physical_entities)
+        {
+          // Assign boundary ID for faces (dim-1)
+          if (physical_dim == dim - 1)
+            entity_to_boundary[{physical_dim, ent}] = physical_tag;
+          if (physical_dim == dim) // Volume materials
+            entity_to_material[{physical_dim, ent}] = physical_tag;
+        }
+    }
+
+  // --- Build boundary face map for all (dim-1) entities ---
+
+  std::map<std::set<unsigned int>, types::boundary_id> boundary_face_map;
+
+  // Process all (dim-1)-dimensional entities for boundary/manifold assignment
+  for (const auto &e : entities)
+    {
+      if (e.first == dim - 1) // Faces in 3D, edges in 2D, vertices in 1D
+        {
+          types::boundary_id boundary_id = numbers::invalid_boundary_id;
+
+          // Check for boundary ID
+          if (auto it = entity_to_boundary.find({e.first, e.second});
+              it != entity_to_boundary.end())
+            boundary_id = it->second;
+
+          if (boundary_id != numbers::invalid_boundary_id)
+            {
+              std::vector<int>                      element_types;
+              std::vector<std::vector<std::size_t>> element_ids, element_nodes;
+
+              // Get all elements on this (dim-1)-entity
+              gmsh::model::mesh::getElements(
+                element_types, element_ids, element_nodes, e.first, e.second);
+
+              for (unsigned int i = 0; i < element_types.size(); ++i)
+                {
+                  if (element_ids[i].empty())
+                    continue;
+
+                  const unsigned int n_nodes_per_elem =
+                    element_nodes[i].size() / element_ids[i].size();
+
+                  for (unsigned int j = 0; j < element_ids[i].size(); ++j)
+                    {
+                      std::set<unsigned int> face_vertices;
```

Comment:
Does this have to be a `set` (also for the outer map holding this), or can it be a more efficient data structure like vector or array?